packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,1130 @@
+   * @returns {string} the value to pass to the input
+   */
+  @autobind
+  private _getVisibleValue(): string {
+    let {
+      value,
+      allowFreeform,
+      autoComplete
+    } = this.props;
+    let {
+      selectedIndex,
+      currentPendingValueValidIndex,
+      currentOptions,
+      currentPendingValue,
+      suggestedDisplayValue
+    } = this.state;
+
+    // If the user passed is a value prop, use that
+    if (value) {
+      return value;
+    }
+
+    let index = selectedIndex;
+
+    if (allowFreeform) {
+
+      // If we are allowing freeform and autocomplete is also true
+      // and we've got a pending value that matches an option, remember
+      // the matched option's index
+      if (autoComplete && this._indexWithinBounds(currentOptions, currentPendingValueValidIndex)) {
+        index = currentPendingValueValidIndex;
+      }
+
+      // Since we are allowing freeform, if there is currently a nonempty pending value, use that
+      // otherwise use the index determined above (falling back to '' if we did not get a valid index)
+      return currentPendingValue !== '' ? currentPendingValue :
+        (this._indexWithinBounds(currentOptions, index) ? currentOptions[index].text : '');
+    } else {
+
+      // If we are not allowing freeform and have a
+      // valid index that matches the pending value,
+      // we know we will need some version of the pending value
+      if (this._indexWithinBounds(currentOptions, currentPendingValueValidIndex)) {
+
+        // If autoComplete is on, return the
+        // raw pending value, otherwise remember
+        // the matched option's index
+        if (autoComplete) {
+          return currentPendingValue;
+        }
+
+        index = currentPendingValueValidIndex;
+      }
+
+      // If we have a valid index then return the text value of that option,
+      // otherwise return the suggestedDisplayValue
+      return this._indexWithinBounds(currentOptions, index) ? currentOptions[index].text : suggestedDisplayValue;
+    }
+  }
+
+  /**
+   * Is the index within the bounds of the array?
+   * @param options - options to check if the index is valid for
+   * @param index - the index to check
+   * @returns {boolean} - true if the index is valid for the given options, false otherwise
+   */
+  @autobind
+  private _indexWithinBounds(options: IComboBoxOption[], index: number): boolean {
+    return index >= 0 && index < options.length;
+  }
+
+  /**
+   * Handler for typing changes on the input
+   * @param updatedValue - the newly changed value
+   */
+  @autobind
+  private _onInputChange(updatedValue: string) {
+    this.props.allowFreeform ?
+      this._processInputChangeWithFreeform(updatedValue) :
+      this._processInputChangeWithoutFreeform(updatedValue);
+  }
+
+  /**
+   * Process the new input's new value when the comboBox
+   * allows freeform entry
+   * @param updatedValue - the input's newly changed value
+   */
+  @autobind
```

Comment:
can you make sure only the methods that need to be bound are decorated with autobind? there is a perf overhead otherwise.